docs/spec_draft.md

```diff
@@ -604,6 +605,86 @@ Afterwards, within each `process_group`:
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.all_reduce
+
+### Semantics
+
+Within each process group in the StableHLO grid, performs reduction, using
+`computations`, over the  values of the `operand` tensor from each process and
+produces a `result` tensor.
+
+The operation splits the StableHLO grid into process groups as follows:
+  * `channel_id <= 0` and `use_global_device_ids = false`,
+    `cross_replica(replica_groups)`.
+  * `channel_id > 0` and `use_global_device_ids = false`,
+    `cross_replica_and_partition(replica_groups)`.
+  * `channel_id > 0` and `use_global_device_ids = true`,
+    `flattened_ids(replica_groups)`.
+
+Afterwards, within each `process_group`:
+  * `operands@receiver = [operand@sender for sender in process_group]` for all
+    `receiver` in `process_group`.
+  * ```
+    result@process[i0, i1, ..., iR-1] =
+        reduce(
+          inputs=[operands@process[j][i0, i1, ..., iR-1] for j in range(size(operands@process))],
+          init_values=[I], # such that `computation` and `I` form a monoid]
+          dimensions=[0],
+          body=computation
+        )
+    ```
```

Comment:
`operand[i0, i1, ..., iR-1]@process[j]` feels better than `operands@process[j][i0, i1, ..., iR-1]`
also operand is singular
`result@process[i0, i1, ..., iR-1]` does not have process id, is that intentional?

Reply:
`j in range(size(operands@process))`
j should iterate in process_group right?

Reply:
>> j should iterate in process_group right?
Once all the operands `operands@process` are collected in a process, the subsequent reduction can be implemented like a local operation. The size is then just size(`operands@process`)


>> operand[i0, i1, ..., iR-1]@process[j] feels better than operands@process[j][i0, i1, ..., iR-1]
also operand is singular

we use the notation `name@process_id` to refer to the value `name` within a particular process.
`operands@process`is what any receiver has collected before starting the reduction. 

`result@process`: process is the process_id here.

Reply:
`operand[i0, i1, ..., iR-1]@process[j]` is an interesting idea, but it doesn't match the definition of `@` which is introduced as: "we use the notation `name@process_id` to refer to the value `name` within a particular process". Let's keep the current version.